src/client/datascience/notebook/intellisense/pythonKernelCompletionProvider.ts

```diff
@@ -274,41 +289,58 @@ export function filterCompletions(
         }
+        // Do nothing for paths and the like inside strings.
+        if (insideString) {
+            return r;
+        }
+
         const wordIndex = word ? r.itemText.indexOf(word) : -1;
+        let newLabel: string | undefined = undefined;
         let newText: string | undefined = undefined;
         let newRange: Range | { inserting: Range; replacing: Range } | undefined = undefined;
 
         // Two cases for filtering. We're at the '.', then the word we have is the beginning of the string.
         // Example, user typed 'df.' and r.itemText is 'df.PassengerId'. Word would be 'df.' in this case.
-        if (word && wordDot && r.itemText.startsWith(word)) {
-            newText = r.itemText.substring(word.length);
+        if (word && wordDot && r.itemText.includes(word)) {
+            newLabel = r.itemText.substring(r.itemText.indexOf(word) + (wordDot ? word.length : 0));
+            newText = r.itemText.substring(r.itemText.indexOf(word) + word.length);
+            const changeInCharacters =
+                (typeof r.label === 'string' ? r.label.length : r.label.label.length) - newText.length;
             newRange =
                 r.range && 'start' in r.range
-                    ? new Range(new Position(r.range.start.line, r.range.start.character + word.length), r.range.end)
+                    ? new Range(
+                          new Position(r.range.start.line, r.range.start.character + changeInCharacters),
+                          r.range.end
+                      )
                     : r.range;
         }
         // We're after the '.' and the user is typing more. We are in the middle of the string then.
         // Example, user typed 'df.Pass' and r.itemText is 'df.PassengerId'. Word would be 'Pass' in this case.
         if (!newText && wordIndex > 0) {
-            newText = r.itemText.substring(wordIndex);
+            newLabel = r.itemText.substring(r.itemText.indexOf(word) + (wordDot ? word.length : 0));
+            newText = r.itemText.substring(r.itemText.indexOf(word) + word.length);
+            const changeInCharacters =
+                (typeof r.label === 'string' ? r.label.length : r.label.label.length) - newText.length;
             newRange =
                 r.range && 'start' in r.range
-                    ? new Range(new Position(r.range.start.line, r.range.start.character + wordIndex), r.range.end)
+                    ? new Range(
+                          new Position(r.range.start.line, r.range.start.character + changeInCharacters),
+                          r.range.end
+                      )
                     : r.range;
         }
-        if (newText && newRange) {
-            return {
-                ...r,
-                sortText: generateSortString(i),
-                label: newText,
-                itemText: newText,
-                range: newRange
-            };
+        if (newLabel && newText && newRange) {
+            r.label = newLabel;
```

Comment:
Label and item to be inserted are different now.
Basically assume you have typed `df.Pass` and the completion item is `Passenger`, then 
* We display the completion item label as `Passenger`
* We add the text to be inserted as `enger` at the position after `s`

Reply:
Previously it worked as we'd replace the entire string, now we just insert at the location where cursor is.